### PR TITLE
[Merged by Bors] - feat: `replaceRec`

### PR DESCRIPTION
**Pull request by fpvandoorn**
* Add a more flexible version of `Expr.replace`

### CODE REVIEW

File: Mathlib/Lean/Expr.lean [hunk 1]
```diff
@@ -61,6 +62,74 @@ def natLit! : Expr → Nat
...
+  expression. If `f? e = none` we proceed to the children of `e`. If
+  `f? e = some (#[e₁, ..., eₙ], g)`, we first recursively apply this function to
+  `#[e₁, ..., eₙ]` to get new expressions `#[f₁, ..., fₙ]`.
+  Then we replace `e` by `g [f₁, ..., fₙ]`.
+
+  Important: In order for this function to terminate, the `[e₁, ..., eₙ]` must all be smaller than
+  `e` according to some measure  (and this measure must also be strictly decreasing on the w.r.t.
+  the structural subterm relation). -/
+@[implementedBy ReplaceRecImpl.replaceUnsafe]
+partial def replaceRec (f? : Expr → Option (Array Expr × (Array Expr → Expr))) (e : Expr) :
```
Comment (gebner) [inline Mathlib/Lean/Expr.lean:126]:
```suggestion
partial def replaceRec (f? : (Expr → Expr) → Expr → Option Expr) (e : Expr) :
```
I wonder if that isn't a more comfortable inferface (implemented with IO.Ref instead of StateM).

File: Mathlib/Tactic/Core.lean [hunk 2]
```diff
@@ -9,6 +9,13 @@ import Mathlib.Lean.Expr
 
 namespace Lean
 
+open Elab
+
+/-- Get the declaration `nm` in the current environment. -/
+def getDecl (nm : Name) : CoreM ConstantInfo := do
```
Comment (gebner) [inline Mathlib/Tactic/Core.lean:15]: This is already in core, called `getConstInfo`.

File: Mathlib/Tactic/ToAdditive.lean [hunk 3]
```diff
@@ -0,0 +1,69 @@
+/-
+Copyright (c) 2017 Mario Carneiro. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Mario Carneiro, Yury Kudryashov, Floris van Doorn
+-/
+import Mathlib.Tactic.Core
+import Mathlib.Lean.Expr
+
```
Comment (gebner) [inline Mathlib/Tactic/ToAdditive.lean:8]: Please put this file in a separate PR.

File: Mathlib/Lean/Expr.lean [hunk 4]
```diff
@@ -61,6 +62,74 @@ def natLit! : Expr → Nat
...
+  Expr :=
+  /- We don't provide a reference implementation, since there is no safe implementation for general
+  `f?`. -/
+  e
```
Comment (gebner) [inline Mathlib/Lean/Expr.lean:130]:
```suggestion
def replaceRec (f? : Expr → Option (Array Expr × (Array Expr → Expr))) (e : Expr) : Expr :=
  unsafe (replaceUnsafeM f? cacheSize e).run' initCache
```

File: Mathlib/Lean/Expr.lean [hunk 5]
```diff
@@ -61,6 +62,74 @@ def natLit! : Expr → Nat
   | lit (Literal.natVal v) _ => v
   | _                        => panic! "nat literal expected"
 
+
+/-! We define a more flexible version of `Expr.replace` where we can use recursive calls even when
```
Comment (gebner) [inline Mathlib/Lean/Expr.lean:66]: Please put this function in a separate file.

File: Mathlib/Lean/Expr.lean (hunk 4 above)
Comment (fpvandoorn) [reply Mathlib/Lean/Expr.lean:130]: I was just following `Expr.replace`. But you're suggesting that the unsafe implementation will be the user-facing one? Does that make all defs that use this also `unsafe`? Or is that the point, since for general `f` this function *is* unsafe?
I can also remove the `@[implementedBy]` under this proposal?

Comment (gebner) [reply Mathlib/Lean/Expr.lean]: I'm just promoting the mathlib-specific `unsafe` *term* syntax.  It replaces the whole `implementedBy` boilerplate.  https://github.com/leanprover-community/mathlib4/pull/125

File: Mathlib/Lean/Expr.lean (hunk 1 above)
Comment (fpvandoorn) [reply Mathlib/Lean/Expr.lean]: That would be a lot nicer interface, but I don't see how to program this. 
* I don't know how to use `IO.Ref` here. Are you suggesting the monad to be `EIO State`? What `IO`-function will be helpful here that `StateM` doesn't provide?
* I'd be also happy with `(f? : Expr → Option ((Expr → Expr) → Expr))`, which might make it easier to implement?
* If `f?` has that type, I can sort-of do it with `StateM`, when replacing the relevant lines to
```lean
    else match f? e with
      | some g => do 
        let s ← get
        let eNew := g λ e => ((visit e).run s).1
        cache i e eNew
```
This has the obvious problem that I don't update the state when computing `eNew`. Can I do it better with `IO`?

Comment (gebner) [reply Mathlib/Lean/Expr.lean]: I was thinking of something like this (but replace the cache type):
```lean
import Std
import Mathlib.Util.TermUnsafe
open Std ShareCommon

unsafe def memoFixImpl {α β : Type} [Inhabited β] (f : (α → β) → (α → β)) (a : α) : β := unsafeBaseIO do
  let cache : IO.Ref (@HashMap Object β ⟨Object.ptrEq⟩ ⟨Object.ptrHash⟩) ← ST.mkRef {}
  let rec fix (a : α) : β := unsafeBaseIO do
    if let some b := (← cache.get).find? (unsafeCast a) then
      return b
    let b := f fix a
    cache.modify (·.insert (unsafeCast a) b)
    b
  fix a

@[implementedBy memoFixImpl]
constant memoFix {α β : Type} [Inhabited β] (f : (α → β) → (α → β)) : α → β

def fibLength : List Nat → Nat :=
  memoFix fun fib xs =>
    dbg_trace xs
    match xs with
    | [] => 0
    | [x] => 1
    | _ :: xs@(_ :: ys) => fib xs + fib ys

-- visits every sublist only once
#eval fibLength [1,2,3,4,5,6,7,8]
```

Comment (fpvandoorn) [reply Mathlib/Lean/Expr.lean]: Ok, I'll try using this. 
I was also wondering about the choice of cache in `Expr.replace`. Obviously `Array` is very quick, but pointer collisions (mod `cacheSize`) seem very expensive, so I was wondering whether a `HashMap`/`RBTree` would be better.

Comment (gebner) [reply Mathlib/Lean/Expr.lean]: > Obviously Array is very quick, but pointer collisions (mod cacheSize) seem very expensive, so I was wondering whether a HashMap/RBTree would be better.

Good question!  I don't really know.  I would imagine that both implementation are good enough for toAdditive, though.

Comment (EdAyers) [reply Mathlib/Lean/Expr.lean]: I've now implemeted Gabriel's suggested signature using a memoFix

File: Mathlib/Lean/Expr.lean (hunk 4 above)
Comment (EdAyers) [reply Mathlib/Lean/Expr.lean]: I think this is moot now that memoFix is being used?

**Comment by EdAyers** [comment]
Requires #237

File: Mathlib/Control/Basic.lean [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+Authors: E.W.Ayers
+-/
+
+/-!
+
+# Basic instances and functions for monads and other structures.
+
+-/
+
+instance : Monad id where
```
Comment (gebner) [inline Mathlib/Control/Basic.lean:13]: BTW, there's `Id` for the type which has a monad instance.

File: Mathlib/Lean/Expr.lean [hunk 7]
```diff
@@ -4,6 +4,7 @@ Released under Apache 2.0 license as described in the file LICENSE.
 Authors: Mario Carneiro, Simon Hudon, Scott Morrison, Keeley Hoek, Robert Y. Lewis, Floris van Doorn
 -/
 import Lean
+import Mathlib.Init.Data.Nat.Basic
```
Comment (gebner) [inline Mathlib/Lean/Expr.lean:7]: Accidental change?

File: Mathlib/Lean/Expr/ReplaceRec.lean [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+import Mathlib.Control.Basic
+import Mathlib.Util.MemoFix
+namespace Lean.Expr
+/-!
+# ReplaceRec
+
+We define a more flexible version of `Expr.replace` where we can use recursive calls even when
+replacing a subexpression. We completely mimic the implementation of `Expr.replace`. -/
+
+private def replaceRecCoreM [Monad M] (f? : (Expr → M Expr) → Expr → M (Option Expr))
```
Comment (gebner) [inline Mathlib/Lean/Expr/ReplaceRec.lean:18]: inline?

File: Mathlib/Lean/Expr/ReplaceRec.lean [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+  `traversal e = some (#[e₁, ..., eₙ], g)`, we first recursively apply this function to
+  `#[e₁, ..., eₙ]` to get new expressions `#[f₁, ..., fₙ]`.
+  Then we replace `e` by `g [f₁, ..., fₙ]`.
+
+  Important: In order for this function to terminate, the `[e₁, ..., eₙ]` must all be smaller than
+  `e` according to some measure  (and this measure must also be strictly decreasing on the w.r.t.
+  the structural subterm relation).
+  -/
+def replaceRecTraversal (traversal : Expr → Option (Array Expr × (Array Expr → Expr))) (e : Expr) : Expr :=
+  e.replaceRecM (M := id) $ fun r e =>
```
Comment (gebner) [inline Mathlib/Lean/Expr/ReplaceRec.lean:52]:
```suggestion
  e.replaceRecM (M := Id) fun r e =>
```

File: test/Expr.lean [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+unsafe def reorderLastArguments : Expr → Expr :=
+Expr.replaceRecTraversal λ e =>
+  let n := e.getAppNumArgs
+  if n ≥ 2 then
+    some (e.getAppArgs, λ es => mkAppN e.getAppFn $ es.swap! (n - 1) (n - 2)) else
+    none
+
+def foo (f : ℕ → ℕ → ℕ) (n₁ n₂ n₃ n₄ : ℕ) : ℕ := f (f n₁ n₂) (f n₃ n₄)
+def bar (f : ℕ → ℕ → ℕ) (n₁ n₂ n₃ n₄ : ℕ) : ℕ := f (f n₄ n₃) (f n₂ n₁)
+#eval (do
```
Comment (gebner) [inline test/Expr.lean:20]:
```suggestion
#eval show TermElabM _ from do
```

File: test/Expr.lean [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+  let d ← getConstInfo `foo
+  let e := d.value!
+  let s ← ppExpr { env := (← getEnv)} e
+  IO.println $ "before: " ++ s
```
Comment (gebner) [inline test/Expr.lean:24]:
```suggestion
  logInfo m!"before {s}"
```

Wonder if this works..

File: Mathlib/Control/Basic.lean (hunk 6 above)
Comment (EdAyers) [reply Mathlib/Control/Basic.lean:13]: ahhhhh

File: test/Expr.lean (hunk 11 above)
Comment (EdAyers) [reply test/Expr.lean]: It does!

File: test/Expr.lean [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+import Mathlib.Init.Data.Nat.Basic
+
+open Lean Meta Elab
+
+section replaceRec
+/-! Test the implementation of `Expr.replaceRec` -/
+
+/-- Reorder the last two arguments of every function in the expression.
+  (The resulting term will generally not be a type-correct) -/
+unsafe def reorderLastArguments : Expr → Expr :=
```
Comment (gebner) [inline test/Expr.lean:11]: Why is this unsafe?

File: Mathlib/Lean/Expr/ReplaceRec.lean [hunk 13]
```diff
@@ -0,0 +1,57 @@
...
+
+`replaceRec f? e` will call `f? r e` where `r = replaceRec f?`.
+If `f? r e = none` then `r` will be called on each immediate subexpression of `e` and reassembled.
+If it is `some x`, traversal terminates and `x` is returned.
+If you wish to recursively replace things in the implementation of `f?`, you can apply `r`.
+
+The function is also memoised, which means that if the
+same expression (by reference) is encountered the cached replacement is used. -/
+def replaceRec (f? : (Expr → Expr) → Expr → Option Expr) (e : Expr) : Expr :=
+replaceRecM (M := Id) f? e
```
Comment (gebner) [inline Mathlib/Lean/Expr/ReplaceRec.lean:38]: Please indent.

File: Mathlib/Lean/Expr/Traverse.lean [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+
+/-!
+# Traversal functions for expressions.
+-/
+
+namespace Lean.Expr
+
+/-- Maps `f` on each immediate child of the given expression. -/
+def traverseChildren [Applicative M] (f : Expr → M Expr) : Expr → M Expr
+| e@(forallE _ d b _) => pure e.updateForallE! <*> f d <*> f b
```
Comment (gebner) [inline Mathlib/Lean/Expr/Traverse.lean:17]: Please indent.

**Review by gebner — COMMENTED** [review]
Please also rename the PR to `feat: replaceRec` (we no longer do the module in parentheses).

File: Mathlib/Lean/Expr/ReplaceRec.lean (hunk 8 above)
Comment (gebner) [reply Mathlib/Lean/Expr/ReplaceRec.lean:18]: Oops, apparently my comment was too cryptic.  My suggestion was to remove the `replaceRecCoreM` definition, and put its body directly in `replaceRecM`.

Comment (EdAyers) [reply Mathlib/Lean/Expr/ReplaceRec.lean]: ahhhh, fixed

File: Mathlib/Lean/Expr/Traverse.lean (hunk 14 above)
Comment (EdAyers) [reply Mathlib/Lean/Expr/Traverse.lean:17]: still getting used to this convention change sorry

**Comment by EdAyers** [comment]
Ok I think that fixes everything

File: Mathlib/Lean/Expr/ReplaceRec.lean [hunk 15]
```diff
@@ -0,0 +1,53 @@
...
+import Mathlib.Util.MemoFix
+namespace Lean.Expr
+/-!
+# ReplaceRec
+
+We define a more flexible version of `Expr.replace` where we can use recursive calls even when
+replacing a subexpression. We completely mimic the implementation of `Expr.replace`. -/
+
+/-- Same as `replaceRec` except over a monad. -/
+def replaceRecM [Monad M] (f? : (Expr → M Expr) → Expr → M (Option Expr)) : Expr → M Expr :=
```
Comment (gebner) [inline Mathlib/Lean/Expr/ReplaceRec.lean:18]: I didn't realize this until now, but `replaceRecM` won't do what you expect it to.  It will cache the `M Expr` and not the `Expr`, so for most monads this won't do any caching at all.  `replaceRec` is fine though.

Comment (EdAyers) [reply Mathlib/Lean/Expr/ReplaceRec.lean:18]: Yeah that's what I thought it was doing and I agree it's pointless

Comment (EdAyers) [reply Mathlib/Lean/Expr/ReplaceRec.lean:18]: but it seemed neater to implement like this.

There is also a method in `src/lean/meta/transform.lean` that does the same kind of thing as this. Maybe this whole PR could be replaced with a call to that.

Comment (EdAyers) [reply Mathlib/Lean/Expr/ReplaceRec.lean:18]: I think that there might be a way of implementing proper monad caching using `[MonadControlT _ M]` but I'm not sure.

Comment (gebner) [reply Mathlib/Lean/Expr/ReplaceRec.lean:18]: In their current state, I'd prefer to remove the `replaceRecM` and co., since they have very misleading performance (and thus should not be used).

To properly implement them, I think we'd need a `memoFixM` combinator.

The `Meta.transform` function is a tradeoff.  It is nicer since it changes the local context, which means you can call `inferType` etc. on the subexpression.  But you can't perform any recursive calls, just change the expression at the current position.  (Which, to be fair, should be enough for `to_additive`.)
